misc/man/rrr_post.1
.Dd 2021-01-14
.TH RRR_POST 1
.SH NAME
Read Route Record is a dynamic module-based message handling program
for collecting, buffering, processing and storing messages and measurements.
.PP
.B rrr_post
parses input from a file or standard input and connects to a UNIX socket created by the main
.B rrr
program to post data to its modules.
.SH SYNOPSIS
.B rrr_post {RRR SOCKET}
.Dl [-f|--file[=]FILENAME|-]
.Dl [-r|--readings[=]reading1,reading2,...]
.Dl [-a|--array-definition[=]ARRAY DEFINITION]
.Dl [-L|--log-delivery]
.Dl [-j|--json]
.Dl [-m|--max-message-size[=]BYTES]
.Dl [-c|--count[=]MAX FILE RECORDS]
.Dl [-t|--topic[=]MQTT TOPIC]
.Dl [-s|--sync]
.Dl [-q|--quiet]
.Dl [-l|--loglevel-translation]
.Dl [-e|--environment-file[=]ENVIRONMENT FILE]
.Dl [-d|--debuglevel[=]FLAGS]
.Dl [-D|--debuglevel-on-exit[=]FLAGS]
.Dl [-h|--help]
.Dl [-v|--version]

.SH DESCRIPTION
This program requires
.Xr rrr(1)
to be running and listening on a UNIX socket with its
.B socket
module. Refer to
.Xr rrr.conf(5)
and
.Xr rrr(1)
on how to set this up.
.B rrr_post
will parse input records in a chosen format and convert them to RRR array messages before sending them on the socket. An input data
record may contain an arbitary number of data fields of different types (strings, numbers etc.).
.PP
.SH OPTIONS
.IP -f|--file[=]FILENAME|-
Read records from a file or standard input (-). The format of the input data must be specified with the
.B -a
option. If left unspecified, no file or standard input reading is performed.
.IP -r|--readings[=]reading1,reading2,...
Generate simple RRR messages, one for each reading specified. The reading value is an unsigned integer stored as 64 bits. If specified, the
reading messages are generated before any reading of a file or standard input begins.
.IP -a|--array-definition[=]DEFINITION
Format of the input data for file or standard input, see the
.B ARRAY DEFINITION
section. If no array is specified, the array "msg" is used (input is RRR messages).
.IP -L|--log-delivery
If given, log messages are expected to be read  which are then delivered as native
RRR log messages to the specified socket.
A socket set up by the log daemon
.B rrr_logd(1)
may be used to receive these log messages.

A given array definition in
.B -a|--array-definition
must provide values with the following fields of which some are optional.

.Bl -tag -width -indent

.It log_level_translated
RFC 5424 loglevel, small integer. If not given, 7 is used.

.It log_flags
If given and set to 1, this indicates that the log message is in JSON format.

.It log_prefix
The log prefix. If not given, "rrr_post" will be used.

.It log_message
The actual log message. This value is mandatory.

.It log_file
.It log_line
The file name and line from where the log message originates, string and integer.
If either is not given, values from the
.B rrr_post
code will be used.

The
.B rrr_logd
daemon is also capable of recieving array messages with the fields above, but the native log messages are more efficient.

.El

If no array definition is specified, the following simple definition expecting
one message per line is used.

.B nsep#log_message,sep1

.IP -j|--json
If given, a flag indicating that the message data is in JSON format is set on any generated log messages.

.IP -m|--max-message-size[=]BYTES
Maximum size of a message when reading. This should be set to prevent messages with missing delimeters to fill up memory.
A value of 0 means unlimited message size.
Defaults to 4096.
.IP -c|--count[=]RECORDS
Stop reading file or standard input after this many records have been read.
.IP -t|--topic[=]MQTT-TOPIC
Set an MQTT topic in the messages generated.
.IP -S|--strip-separators
Strip out any parsed
.B sep
fields from arrays.
.IP -s|--sync
If array parsing fails (invalid input data), try to parse again after skipping one byte forward in the input.
.IP -q|--quiet
Don't print statistics upon exit.
.IP -d|--debuglevel[=]FLAGS
Debuglevel to use.
.IP -D|--debuglevel-on-exit[=]FLAGS
Debuglevel to use when program exits.
.IP -l|--loglevel-translation
Translate output loglevels to RFC5424 format. Internal loglevels will be translated so that error messages become ERROR
and other messages DEBUG (loglevels 3 and 7).
.IP -e|--environment-file
See
.Xr rrr(1)
for usage.
.PP
.SH SIGNALS
.IP SIGURS1
Print sent message statistics, then continue.
.IP SIGINT/CTRL+C
Complete processing of current message then exit. SIGINT a second time causes immediate exit.
.IP SIGTERM
Exit immediately.
.IP SIGPIPE
Ignored (an error message is printed).
.PP
.SH ARRAY DEFINITION
A range of data types may be used in the data records given to
.B rrr_post
. A record consists of one or more import items of different types.
Some types are of fixed length, and others require delimeters. An array definition describes the data to expect 
in each record.
.PP
Array definition may or may not contain conditional branches.
This allows different input values to be parsed based on previously parsed values.
Coditional branching is explained in the next chapter.
.PP
It is important to make sure that it is always possible to figure out the lenght of each received record from the data it contains.
For instance, a definition cannot end with a number given as ASCII string as it's not possible to determin when it ends.
Such data types are called
.B weak dynamic lenght types,
and these must be delimeted by a separator character which is guaranteed not to appear inside the data.
RRR does not smart checking or validation of definition, but parsing of input data will fail if value boundaries cannot be correctly identified.
.PP
If parsing fails, this might lead to for instance a TCP connection to be closed or queued up UDP data to be dropped.
Using
.B debuglevel 4
will expose how input data are parsed using the array definitions, this might assist in debugging such problems.
.PP
An array definition consists of one or more
.B positions
which each specifies a particular type of data to parse.
Sequential fixed length values of the same type may be grouped together in a single position, and each item may be tagged to
ease the extraction of data in later processing.
Non-dynamic types may have both a length or element count set.
Parsing four elements of one byte and parsing one element of four bytes only differ in how the data is represented after the array has been parsed.    
.SS SYNTAX
type1[length1][s|u|][@count1][#tag1][,type2[length2][s|u|][@count2][#tag2]][,...]
.PP
.Bl -tag -width -indent
.It type
Identifier name of the type

.It length
Length in bytes of the type (if required)

.It count
Item count of the specific type, defaults to 1 if not specified

.It tag
Optional custom identifier tag of the type (containing letters, numbers and/or underscores)
.El
.SS FIXED LENGTH TYPES
These types require the
.I length
field to be specified.
.PP
.Bl -tag -width -indent
.It be
Unsigned number in big endian binary format. Length must be in the range 1-8.

.It le
Unsigned number in little endian binary format. Length must be in the range 1-8.

.It h
Unsigned number in the endianess of the machine. Might be unsafe for network transfer. Length must be in the range 1-8.

.It blob
Arbitary binary data. Lengt must be in the range 1-1024.

.It sep
One or more separator characters. Matches ! " # $ % & ' ( ) * + , - . / : ; < = > ? @ [ \ ] ^ _ ` { | } ~ NULL ETX EOH LF CR TAB.
Length must be in the range 1-64.

.It stx
One or more STX or SOH characters.
Length must be in the range 1-64.
.El
.PP
Types
.B be, le
and
.B h
may be followed by an
.B s
after the length specifier to indicate that the input number is signed. If instead
.B u
or nothing is set here, the value is treated as unsigned. No other types may have sign flag set.
.SS WEAK DYNAMIC LENGTH TYPES
The length of these types are identified automatically and must not have length set. They cannot be at the end of a definition.
.PP
.Bl -tag -width -indent
.It ustr
An unsigned integer encoded with ASCII characters 0-9. Stored with 64-bits. Number may begin with one or more white spaces.

.It istr
A signed integer encoded with ASCII characters 0-9 optionally preceeded by - or +. Stored with 64-bits. Number may begin with one or more white spaces.

.It fixp
The RRR fixed decimal type encoded with ASCII characters 0-9 (and A-F). May include a single dot . to separate integer from fraction,
and the integer part may be preceded with a sign (- or +). Stored with 64-bits where 1 bit is the sign, 39 bits is the integer and
24 bits are the fraction. May be preceeded with 10# or 16# to indicate use of base 10 or base 16 conversion, default is base 10. 
.El
.SS STRONG DYNAMIC LENGTH TYPES

debian/changelog
rrr (1.36-15) bookworm; urgency=medium

  * RUNNING DEVELOPMENT RELEASE

  * Features:
    * New methods on JS Message object: data_as_object and data_as_utf8
    * Support for Node 23
    * Support for Debian Trixie
    * Log daemon may start a wrapper process for further processing of
      messages prior to them being printed.
    * Configurable sub instances in cmodules
    * Native JSON logging from log daemon

  * Fixes:
    * BAIL in method and route definitions not working correctly, caused
      program exit when triggered
    * FALSE value in method and route definitions not working, TRUE was pushed
      instead
    * Missing check for overshoot and undershoot while fading in ArtNet
    * Incorrect counting of FIFO buffer entries in certain situations, could
      possibly result in buffers over time being reported as full when they
      were not. The problem only affected counting of retained messages in MQTT
      broker.
    * Misc. compilation problems using libc++ on FreeBSD, which is required
      then V8 is used on this platform
    * MQTT client not always reconnecting as configured but restarted
    * JS incorrectly loading the same module multiple times if included
      from different scripts

 -- Atle Solbakken <atle@goliathdns.no>  Mon, 24 Mar 2025 12:04:31 +0100

rrr (1.35-1) bookworm; urgency=medium

  * Features:
    * Configuration option in http client to trap any header field

  * Fixes:
    * Crashes in MQTT for malformed packets with tailing payload
    * Bad MQTT CONNECT packet assembling for V5 when will flag was not set
    * Log messages to stats engine  getting lost due to events delivered to
      incorrect event queue
    * Memory leak in Python3 where values added to array had incorrect
      reference count causing memory leaks
    * Broken parsing of leading spaces in istr and ustr types in arrays
    * Crash in array tree import for division overflow in conditions,
      emit soft error instead.

 -- Atle Solbakken <atle@goliathdns.no>  Wed, 27 Nov 2024 18:32:12 +0100

rrr (1.33-1) bookworm; urgency=medium

  * Features:
    * Log socket daemon with ability to collect logs from multiple RRR
      processes

  * Improvements:
    * Remove all references to http2 when RRR is built without nghttp2, in
      particular avoid noisy notification caused by inconsistencies in the
      code
    * Speed up tests by running them in parallel
    * In httpclient, by default ignore all meta fields from httpserver which
      should usually not be added as request fields.
    * In httpclient, prevent using // endpoint when using endpoint from topic
      and url is /

  * Fixes:
    * Remove stale python directory from source causing build to fail in some
      cases
    * Misc. hangs and crashes when parsing invalid configuration files
    * Misc. crashes in HTTP handling

 -- Atle Solbakken <atle@goliathdns.no>  Fri, 01 Nov 2024 07:02:09 +0100

rrr (1.31-5) bookworm; urgency=medium

  * Fixes:
    * Preflight headers not sendt for OPTIONS request
    * Move setproctitle() call to ensure that it cannot clear environment
      variables prior to them being accessed during initialization.

 -- Atle Solbakken <atle@goliathdns.no>  Mon, 12 Aug 2024 13:05:28 +0200

rrr (1.31-1) bookworm; urgency=medium

  * Features:
    * New option http_server_decode_endpoint to decode the endpoint field
      when receiving full requests.
    * Support HTTP3 over QUIC in HTTP server and client.
    * Start RRR in single thread mode which is useful for debugging and
      on slower hardware.
    * Support for more ModBus functions
      - By Fredrik Rosenqvist <rosenqvistfredrik@hotmail.com>
    * New method rrr_array.has() in the Python module to check if an array tag
      exists.
      - By Fredrik Rosenqvist <rosenqvistfredrik@hotmail.com>
    * New option http_server_topic_format to include more information about a
      request in the topic of messages generated by HTTP server module.
      - By Fredrik Rosenqvist <rosenqvistfredrik@hotmail.com>

  * Improvements:
    * Implement polite close of connections during shutdown for HTTP1
      and HTTP2 in HTTP server.
    * Optimization in message database to avoid excessive chdir calls.

  * Fixes:
    * Broken TLS mode in MQTT client.

src/rrr_post.c
/*

Read Route Record

Copyright (C) 2019-2024 Atle Solbakken atle@goliathdns.no

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>

#include "lib/util/macro_utils.h"
#include "main.h"
#include "../build_timestamp.h"
#include "lib/rrr_config.h"
#include "lib/log.h"
#include "lib/allocator.h"
#include "lib/version.h"
#include "lib/cmdlineparser/cmdline.h"
#include "lib/array.h"
#include "lib/array_tree.h"
#include "lib/socket/rrr_socket.h"
#include "lib/socket/rrr_socket_read.h"
#include "lib/socket/rrr_socket_common.h"
#include "lib/rrr_strerror.h"
#include "lib/read.h"
#include "lib/messages/msg_msg.h"
#include "lib/messages/msg_log.h"
#include "lib/util/rrr_time.h"
#include "lib/util/gnu.h"
#include "lib/util/linked_list.h"
#include "lib/helpers/log_helper.h"

RRR_CONFIG_DEFINE_DEFAULT_LOG_PREFIX("rrr_post");

#define RRR_POST_DEFAULT_ARRAY_DEFINITION       "msg"
#define RRR_POST_DEFAULT_MAX_MESSAGE_SIZE       4096
#define RRR_POST_DEFAULT_LOG_ARRAY_DEFINITION   "nsep#log_message,sep1"

static volatile int rrr_post_abort = 0;
static volatile int rrr_post_print_stats = 0;

static const struct cmd_arg_rule cmd_rules[] = {
        {CMD_ARG_FLAG_NO_FLAG,        '\0',    "socket",               "{RRR SOCKET}"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'f',    "file",                 "[-f|--file[=]FILENAME|-]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT |
         CMD_ARG_FLAG_SPLIT_COMMA,     'r',    "readings",             "[-r|--readings[=]reading1,reading2,...]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'a',    "array-definition",     "[-a|--array-definition[=]ARRAY DEFINITION]"},
	{0,                            'L',    "log-delivery",         "[-L|--log-delivery]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'm',    "max-message-size",     "[-m|--max-message-size]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'c',    "count",                "[-c|--count[=]MAX FILE ELEMENTS]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    't',    "topic",                "[-t|--topic[=]MQTT TOPIC]"},
        {0,                            's',    "sync",                 "[-s|--sync]"},
        {0,                            'S',    "strip-separators",     "[-S|--strip-separators]"},
        {0,                            'q',    "quiet",                "[-q|--quiet]"},
        {0,                            'l',    "loglevel-translation", "[-l|--loglevel-translation]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'e',    "environment-file",     "[-e|--environment-file[=]ENVIRONMENT FILE]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'd',    "debuglevel",           "[-d|--debuglevel[=]DEBUG FLAGS]"},
        {CMD_ARG_FLAG_HAS_ARGUMENT,    'D',    "debuglevel-on-exit",   "[-D|--debuglevel-on-exit[=]DEBUG FLAGS]"},
        {0,                            'h',    "help",                 "[-h|--help]"},
        {0,                            'v',    "version",              "[-v|--version]"},
        {0,                            '\0',    NULL,                   NULL}
};

struct rrr_post_reading {
	RRR_LL_NODE(struct rrr_post_reading);
	uint64_t value;
};

struct rrr_post_reading_collection {
	RRR_LL_HEAD(struct rrr_post_reading);
};

struct rrr_post_data {
	char *filename;
	char *socket_path;
	char *topic;
	struct rrr_post_reading_collection readings;
	uint64_t max_elements;
	uint64_t max_message_size;
	uint64_t elements_count;
	struct rrr_array_tree *tree;

	int strip_separators;
	int sync_byte_by_byte;
	int quiet;
	int log_delivery;

	int input_fd;
	int output_fd;

	uint64_t start_time;
};

static void __rrr_post_signal_handler (int s) {
	if (s == SIGUSR1) {
		rrr_post_print_stats = 1;
	}
	else if (s == SIGPIPE) {
		RRR_MSG_0("Received SIGPIPE, ignoring\n");
	}
	else if (s == SIGTERM) {
		RRR_MSG_0("Received SIGTERM, exiting\n");
		exit(EXIT_FAILURE);
	}
	else if (s == SIGINT) {
		// Allow double ctrl+c to close program immediately
		signal(SIGINT, SIG_DFL);
		rrr_post_abort = 1;
	}
}

static void __rrr_post_data_init (struct rrr_post_data *data) {
	memset (data, '\0', sizeof(*data));
	data->start_time = rrr_time_get_64();
}

static void __rrr_post_destroy_data (struct rrr_post_data *data) {
	RRR_FREE_IF_NOT_NULL(data->filename);
	RRR_FREE_IF_NOT_NULL(data->socket_path);
	RRR_FREE_IF_NOT_NULL(data->topic);
	RRR_LL_DESTROY(&data->readings, struct rrr_post_reading, rrr_free(node));
	if (data->tree != NULL) {
		rrr_array_tree_destroy(data->tree);
	}
}

static int __rrr_post_add_readings (struct rrr_post_data *data, struct cmd_data *cmd) {
	for (cmd_arg_count i = 0; 1; i++) {
		const char *reading = cmd_get_value(cmd, "readings", i);
		if (reading != NULL) {
			for (cmd_arg_count j = 0; 1; j++) {
				reading = cmd_get_subvalue(cmd, "readings", i, j);
				if (reading != NULL) {
					uint64_t value;
					if (cmd_convert_uint64_10(reading, &value) != 0) {
						RRR_MSG_0("Error in reading '%s', not an unsigned integer\n", reading);
						return 1;
					}
					struct rrr_post_reading *reading_new = rrr_allocate(sizeof(*reading_new));
					if (reading_new == NULL) {
						RRR_MSG_0("Could not allocate memory in __rrr_post_add_readings\n");
						return 1;
					}
					reading_new->value = value;
					RRR_LL_APPEND(&data->readings, reading_new);
				}
				else {
					break;
				}
			}
		}
		else {
			break;
		}
	}
	return 0;
}

static int __rrr_post_parse_config (struct rrr_post_data *data, struct cmd_data *cmd) {
	int ret = 0;

	char *array_tree_tmp = NULL;

	// Socket
	const char *socket = cmd_get_value(cmd, "socket", 0);
	if (socket == NULL || *socket == '\0') {
		RRR_MSG_0("No socket path specified\n");
		ret = 1;
		goto out;
	}

	data->socket_path = rrr_strdup(socket);
	if (data->socket_path == NULL) {
		RRR_MSG_0("Could not allocate memory in __rrr_post_parse_config\n");
		ret = 1;
		goto out;
	}

	// Strip separators
	if (cmd_exists(cmd, "strip-separators", 0)) {
		data->strip_separators = 1;
	}
	else {
		data->strip_separators = 0;
	}

	// Sync byte by byte
	if (cmd_exists(cmd, "sync", 0)) {
		data->sync_byte_by_byte = 1;
	}
	else {
		data->sync_byte_by_byte = 0;
	}

	// Quiet operation
	if (cmd_exists(cmd, "quiet", 0)) {
		data->quiet = 1;
	}
	else {
		data->quiet = 0;
	}

	// Filename
	const char *filename = cmd_get_value(cmd, "file", 0);
	if (cmd_get_value (cmd, "file", 1) != NULL) {
		RRR_MSG_0("Error: Only one filename argument may be specified\n");
		ret = 1;
		goto out;
	}
	if (filename != NULL) {
		data->filename = rrr_strdup(filename);
		if (data->filename == NULL) {
			RRR_MSG_0("Could not allocate memory in __rrr_post_parse_config\n");
			ret = 1;
			goto out;
		}
	}

	const char *topic = cmd_get_value(cmd, "topic", 0);
	if (cmd_get_value (cmd, "topic", 1) != NULL) {
		RRR_MSG_0("Error: Only one topic argument may be specified\n");
		ret = 1;
		goto out;
	}
	if (topic != NULL) {
		if (strlen(topic) > RRR_MSG_TOPIC_MAX) {
			RRR_MSG_0("Error: Specified topic too long (%llu>%llu)\n",
				(long long unsigned) strlen(topic),
				(long long unsigned) RRR_MSG_TOPIC_MAX);
			ret = 1;
			goto out;
		}
		data->topic = rrr_strdup(topic);
		if (data->topic == NULL) {
			RRR_MSG_0("Could not allocate memory in __rrr_post_parse_config\n");
			ret = 1;
			goto out;
		}
	}

	// Readings
	if ((ret = __rrr_post_add_readings(data, cmd)) != 0) {
		goto out;
	}

	// Count
	const char *max_elements = cmd_get_value(cmd, "count", 0);
	if (cmd_get_value (cmd, "count", 1) != NULL) {
		RRR_MSG_0("Error: Only one 'count' argument may be specified\n");
		ret = 1;
		goto out;
	}
	if (max_elements != NULL) {
		if (cmd_convert_uint64_10(max_elements, &data->max_elements)) {
			RRR_MSG_0("Could not understand argument 'count', must be and unsigned integer\n");
			ret = 1;
			goto out;
		}
	}

	// Log delivery
	if (cmd_exists(cmd, "log-delivery", 0)) {
		data->log_delivery = 1;
	}

	// Array definition
	const char *array_definition = cmd_get_value(cmd, "array-definition", 0);

	if (array_definition == NULL || *array_definition == '\0') {
		array_definition = data->log_delivery
			? RRR_POST_DEFAULT_LOG_ARRAY_DEFINITION
			: RRR_POST_DEFAULT_ARRAY_DEFINITION;
	}

	array_tree_tmp = rrr_allocate(strlen(array_definition) + 1 + 1); // plus extra ; plus \0
	if (array_tree_tmp == NULL) {
		RRR_MSG_0("Could not allocate temporary arry tree string in parse_config\n");
		ret = 1;
		goto out;
	}

	sprintf(array_tree_tmp, "%s;", array_definition);

	if (rrr_array_tree_interpret_raw (
			&data->tree,
			array_tree_tmp,
			rrr_length_from_size_t_bug_const(strlen(array_tree_tmp)),
			"-"
	) != 0 || data->tree == NULL) {
		RRR_MSG_0("Error while parsing array tree definition\n");
		ret = 1;
		goto out;
	}

	if (cmd_get_value (cmd, "array-definition", 1) != NULL) {
		RRR_MSG_0("Error: Only one array_definition argument may be specified\n");
		ret = 1;
		goto out;
	}

	// Max message size, make sure default value is being set
	data->max_message_size = RRR_POST_DEFAULT_MAX_MESSAGE_SIZE;

	const char *max_message_size = cmd_get_value(cmd, "max-message-size", 0);
	if (cmd_get_value (cmd, "max-message-size", 1) != NULL) {
		RRR_MSG_0("Error: Only one 'max-message-size' argument may be specified\n");
		ret = 1;
		goto out;
	}
	if (max_message_size != NULL) {
		if (cmd_convert_uint64_10(max_message_size, &data->max_message_size)) {
			RRR_MSG_0("Could not understand argument 'max-message-size', must be and unsigned integer\n");
			ret = 1;
			goto out;
		}
	}

	out:
	RRR_FREE_IF_NOT_NULL(array_tree_tmp);
	return ret;
}

static int __rrr_post_connect(struct rrr_post_data *data) {
	int ret = 0;

	if (rrr_socket_unix_connect(&data->output_fd, "rrr_post", data->socket_path, 0) != RRR_SOCKET_OK) {
		RRR_MSG_0("Could not connect to socket\n");
		ret = 1;
		goto out;
	}

	out:
	return ret;
}

static int __rrr_post_open(struct rrr_post_data *data) {
	int ret = 0;

	if (data->filename == NULL) {
		data->input_fd = -1;
	}
	else if (strcmp(data->filename, "-") == 0) {
		data->input_fd = STDIN_FILENO;
	}
	else {
		data->input_fd = rrr_socket_open(data->filename, O_RDONLY, 0, "rrr_post", 0);
		if (data->input_fd < 0) {
			RRR_MSG_0("Could not open input file %s: %s\n",
					data->filename, rrr_strerror(errno));
			ret = 1;
			goto out;
		}
	}

	out:
	return ret;
}

static void __rrr_post_close(struct rrr_post_data *data) {
	if (data->output_fd > 0) {
		rrr_socket_close(data->output_fd);
	}
	if (data->input_fd > 0) {
		rrr_socket_close(data->input_fd);
	}
}

static int __rrr_post_send_message(struct rrr_post_data *data, struct rrr_msg *message) {
	int ret = 0;

	if ((ret = rrr_socket_common_prepare_and_send_msg_blocking (
			message,
			data->output_fd,
			NULL,
			NULL,
			NULL,
			0 /* Not silent */
	)) != 0) {
		RRR_MSG_0("Error while sending message in __rrr_post_send_message\n");
		goto out;
	}

	out:
	return ret;
}

static int __rrr_post_send_reading(struct rrr_post_data *data, struct rrr_post_reading *reading) {
	int ret = 0;

	struct rrr_msg_msg *message = NULL;
	char *text = NULL;

	if (rrr_asprintf(&text, "%" PRIu64, reading->value) <= 0) {
		RRR_MSG_0("Could not create reading text in __rrr_post_send_reading\n");
		ret = 1;
		goto out;
	}

	if (rrr_msg_msg_new_empty (
			&message,
			MSG_TYPE_MSG,
			MSG_CLASS_DATA,
			rrr_time_get_64(),
			0,
			rrr_length_from_size_t_bug_const(strlen(text) + 1)
	)) {
		RRR_MSG_0("Could not allocate message in __rrr_post_send_reading\n");
		ret = 1;
		goto out;
	}

	memcpy(MSG_DATA_PTR(message), text, strlen(text) + 1);

	ret = __rrr_post_send_message(data, (struct rrr_msg *) message);

	out:
	RRR_FREE_IF_NOT_NULL(text);
	RRR_FREE_IF_NOT_NULL(message);
	return ret;
}

static int __rrr_post_send_readings(struct rrr_post_data *data) {
	int ret = 0;

	RRR_LL_ITERATE_BEGIN(&data->readings, struct rrr_post_reading);
		if ((ret = __rrr_post_send_reading(data, node)) != 0) {
			goto out;
		}
	RRR_LL_ITERATE_END();

	out:
	return ret;
}
struct rrr_post_read_callback_data {
	struct rrr_post_data *data;
};

static int __rrr_post_read_log_delivery_callback (
		struct rrr_read_session *read_session,
		struct rrr_array *array_final,
		void *arg
) {
	struct rrr_post_read_callback_data *callback_data = arg;
	struct rrr_post_data *data = callback_data->data;

	(void)(read_session);
	(void)(data);

	int ret = 0;

	struct rrr_msg_log *msg_log = NULL;
	char *log_message = NULL;
	char *log_prefix = NULL;
	char *log_file = NULL;
	uint8_t log_level_translated = 7;
	int log_line = 0;

	if ((ret = rrr_log_helper_extract_log_fields_from_array (
			&log_file,
			&log_line,
			&log_level_translated,
			&log_prefix,
			&log_message,
			array_final
	)) != 0) {
		RRR_MSG_0("Failed to extract log fields\n");
		goto out;
	}

	int no_file_or_line = log_file == NULL || *log_file == '\0' || log_line == 0;
	int no_prefix       = log_prefix == NULL || *log_prefix == '\0';

	if ((ret = rrr_msg_msg_log_new (
			&msg_log,
			no_file_or_line ? __FILE__ : log_file,
			no_file_or_line ? __LINE__ : log_line,
			log_level_translated,
			RRR_MSG_LOG_LEVEL_ORIG_NOT_GIVEN,
			no_prefix ? "rrr_post" : log_prefix,
			log_message
	)) != 0) {
		goto out;
	}

	if ((ret = __rrr_post_send_message(data, (struct rrr_msg *) msg_log)) != 0) {
		// Message printed in called function
	}

	out:
	RRR_FREE_IF_NOT_NULL(msg_log);
	RRR_FREE_IF_NOT_NULL(log_message);
	RRR_FREE_IF_NOT_NULL(log_prefix);
	RRR_FREE_IF_NOT_NULL(log_file);
	return ret;
}

static int __rrr_post_read_array_callback (
		struct rrr_read_session *read_session,
		struct rrr_array *array_final,
		void *arg
) {
	struct rrr_post_read_callback_data *callback_data = arg;
	struct rrr_post_data *data = callback_data->data;

	(void)(read_session);

	int ret = 0;

	struct rrr_msg_msg *message = NULL;

	if (data->strip_separators) {
		rrr_array_strip_type(array_final, &rrr_type_definition_sep);
	}

	if ((ret = rrr_array_new_message_from_array (
			&message,
			array_final,
			rrr_time_get_64(),
			data->topic,
			(data->topic != NULL ? rrr_u16_from_biglength_bug_const(strlen(data->topic)) : 0)
	)) != 0) {
		RRR_MSG_0("Could not create or send message in __rrr_post_read_callback\n");
		goto out;
	}

	if ((ret = __rrr_post_send_message(data, (struct rrr_msg *) message)) != 0) {
		// Message printed in called function
	}

	data->elements_count++;

	out:
	RRR_FREE_IF_NOT_NULL(message);

	return ret;
}

static void __rrr_post_print_statistics (struct rrr_post_data *data) {
	uint64_t runtime = rrr_time_get_64() - data->start_time;
	runtime = runtime / 1000 / 1000;

	if (runtime == 0) {
		runtime = 1;
	}

	uint64_t speed = data->elements_count / runtime;

	RRR_DBG("Processed messages: %" PRIu64 " (%" PRIu64 " m/s), limit: %" PRIu64 ", run time: %" PRIu64 "\n",
			data->elements_count,
			speed,
			data->max_elements,
			runtime
	);
}

static int __rrr_post_read (struct rrr_post_data *data) {
	int ret = 0;

	struct rrr_array array_tmp = {0};

	struct rrr_read_session_collection read_sessions;
	rrr_read_session_collection_init(&read_sessions);

	if (data->filename == NULL) {
		goto out;
	}

	int socket_read_flags = RRR_SOCKET_READ_METHOD_READ_FILE | RRR_SOCKET_READ_USE_TIMEOUT | RRR_SOCKET_READ_NO_GETSOCKOPTS;
	if (data->max_elements == 0 && strcmp (data->filename, "-") != 0) {
		socket_read_flags |= RRR_SOCKET_READ_CHECK_EOF;
	}
	if (strcmp (data->filename, "-") == 0) {
		socket_read_flags |= RRR_SOCKET_READ_CHECK_POLLHUP;
	}

	struct rrr_post_read_callback_data callback_data = {
			data
	};
	while (	ret == 0 &&
			(data->max_elements == 0 || data->elements_count < data->max_elements) &&
			rrr_post_abort == 0
	) {
		uint64_t bytes_read = 0;
		ret = rrr_socket_common_receive_array_tree (
				&bytes_read,
				&read_sessions,
				data->input_fd,
				socket_read_flags,
				&array_tmp,
				data->tree,
				data->sync_byte_by_byte,
				4096,
				0, // No ratelimit interval
				0, // No ratelimit max bytes
				data->max_message_size > RRR_LENGTH_MAX
					? RRR_LENGTH_MAX
					: (rrr_length) data->max_message_size,
				data->log_delivery
					? __rrr_post_read_log_delivery_callback
					: __rrr_post_read_array_callback,
				NULL,
				&callback_data
		);

		/*
		 * XXX [atle]: Removed as it might cause infinite loop if invalid
		 *             data is received. If something breaks because this
		 *             is commented out, another solution should be found.
		 *
		if (ret == RRR_SOCKET_SOFT_ERROR) {
			RRR_MSG_0("Warning: Invalid or unexpected data received\n");
			ret = 0;
		}
		*/

		if (rrr_post_print_stats != 0) {
			__rrr_post_print_statistics(data);
			rrr_post_print_stats = 0;
		}
	}

	if (ret == RRR_SOCKET_READ_EOF) {
		RRR_DBG_1("End of file reached\n");
		ret = 0;
	}

	out:
	rrr_array_clear(&array_tmp);
	rrr_read_session_collection_clear(&read_sessions);
	return ret;
}

int main (int argc, const char **argv, const char **env) {
	if (!rrr_verify_library_build_timestamp(RRR_BUILD_TIMESTAMP)) {
		fprintf(stderr, "Library build version mismatch.\n");
		exit(EXIT_FAILURE);
	}

	int ret = EXIT_SUCCESS;

	if (rrr_allocator_init() != 0) {
		ret = EXIT_FAILURE;
		goto out_final;
	}
	if (rrr_log_init() != 0) {
		ret = EXIT_FAILURE;
		goto out_cleanup_allocator;
	}
	rrr_strerror_init();

	struct cmd_data cmd;
	struct rrr_post_data data;

	cmd_init(&cmd, cmd_rules, argc, argv);
	__rrr_post_data_init(&data);

	if ((ret = rrr_main_parse_cmd_arguments_and_env(&cmd, env, CMD_CONFIG_DEFAULTS)) != 0) {
		goto out;
	}

	if (rrr_main_print_banner_help_and_version(&cmd, 2) != 0) {
		goto out;
	}

	if ((ret = __rrr_post_parse_config(&data, &cmd)) != 0) {
		goto out;
	}

	// Connect to RRR socket
	if ((ret = __rrr_post_connect(&data)) != 0) {
		goto out;
	}

	// Send readings defined in command arguments
	if ((ret = __rrr_post_send_readings(&data)) != 0) {
		goto out;
	}

	// Open input file if defined
	if ((ret = __rrr_post_open(&data)) != 0) {
		goto out;
	}

	struct sigaction action;
	action.sa_handler = __rrr_post_signal_handler;
	sigemptyset (&action.sa_mask);
	action.sa_flags = 0;

	// We generally ignore sigpipe and use NONBLOCK on all sockets
	sigaction (SIGPIPE, &action, NULL);
	// Used to set rrr_post_abort = 1. The signal is set to default afterwards
	// so that a second SIGINT will terminate the process
	sigaction (SIGINT, &action, NULL);
	// Used to print statistics
	sigaction (SIGUSR1, &action, NULL);
	// Exit immediately with EXIT_FAILURE
	sigaction (SIGTERM, &action, NULL);

	// Send readings from input file or stdin
	if ((ret = __rrr_post_read(&data)) != 0) {
		goto out;
	}

	out:
		if (data.quiet == 0) {
			__rrr_post_print_statistics(&data);
		}

		rrr_config_set_debuglevel_on_exit();
		__rrr_post_close(&data);
		__rrr_post_destroy_data(&data);
		cmd_destroy(&cmd);
		rrr_socket_close_all();
		rrr_strerror_cleanup();
		rrr_log_cleanup();
	out_cleanup_allocator:
		rrr_allocator_cleanup();
	out_final:
		return ret;
}
